pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java

```diff
@@ -45,20 +46,36 @@ public TypeFactory(RelDataTypeSystem typeSystem) {
 
   public RelDataType createRelDataTypeFromSchema(Schema schema) {
     Builder builder = new Builder(this);
+    Predicate<FieldSpec> isNullable;
+    if (schema.getOptions().getNullHandling().supportsV2()) {
+      isNullable = schema.getOptions().getNullHandling()::isNullable;
+    } else {
+      isNullable = fieldSpec -> false;
+    }
     for (Map.Entry<String, FieldSpec> e : schema.getFieldSpecMap().entrySet()) {
-      builder.add(e.getKey(), toRelDataType(e.getValue()));
+      builder.add(e.getKey(), toRelDataType(e.getValue(), isNullable));
     }
     return builder.build();
   }
 
-  private RelDataType toRelDataType(FieldSpec fieldSpec) {
+  private RelDataType toRelDataType(FieldSpec fieldSpec, Predicate<FieldSpec> isNullable) {
+    RelDataType type = createSqlType(getSqlTypeName(fieldSpec));
+    boolean isArray = !fieldSpec.isSingleValueField();
+    if (isArray) {
+      type = createArrayType(type, -1);
+    }
+    if (isNullable.test(fieldSpec)) {
+      type = createTypeWithNullability(type, true);
+    }
```

Comment:
do we consider `isNullable` flag here refer to the array itself or this is refering to the element inside the array can be null? 
it is interesting b/c some DB consider array to be null <=> empty; but postgres do not (e.g. can have null array)
let's make sure we document it properly here.

Reply:
Most databases do not correctly support arrays (ie mysql 6.x). Postgres, on the other hand, has a sane type system (that even supports inheritance!). AFAIK when a column of type array is nullable it means the array itself is nullable. I don't know if each element in the array can be tuned nullable or not.

Reply:
yeah agreed. as long as we document & best effort align with postgers it should be good :-D